lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2806,100 +2813,127 @@ migrationPlanToSelectionWithdrawals plan rewardWithdrawal outputAddressesToCycle
-            liftIO $ stakeKeyDeposit <$> currentProtocolParameters nl
+calcMinimumDeposit DBLayer{..}  netLayer wid =
+    throwInIO $ mapExceptT atomically (isStakeKeyRegistered wid) >>= \case
+        True -> pure $ Coin 0
+        False -> liftIO $ stakeKeyDeposit <$> currentProtocolParameters netLayer
   where
-    throwInIO :: ExceptT ErrNoSuchWallet IO a -> IO a
-    throwInIO x = runExceptT x >>= \case
-        Right a -> pure a
-        Left e -> throwIO $ ExceptionNoSuchWallet e
+    throwInIO :: forall a. ExceptT ErrNoSuchWallet IO a -> IO a
+    throwInIO = runExceptT >=> either (throwIO . ExceptionNoSuchWallet) pure
 
-    db = ctx ^. dbLayer @IO @s @k
-    nl = ctx ^. networkLayer
+data DelegationFee = DelegationFee
+    { feePercentiles :: (Percentile 10 Fee, Percentile 90 Fee)
+    , deposit :: Coin
+    }
 
--- | Estimate the transaction fee for a given coin selection algorithm by
--- repeatedly running it (100 times) and collecting the results. In the returned
--- 'FeeEstimation', the minimum fee is that which 90% of the sampled fees are
--- greater than. The maximum fee is the highest fee observed in the samples.
-estimateFee
-    :: forall m. Monad m
-    => ExceptT ErrSelectAssets m Coin
-    -> ExceptT ErrSelectAssets m FeeEstimation
-estimateFee
+delegationFee
+    :: forall s k (n :: NetworkDiscriminant)
+     . ( AddressBookIso s
+       , s ~ SeqState n k
+       , BoundedAddressLength k
+       , Typeable k
+       , Typeable n
+       )
+    => DBLayer IO s k
+    -> NetworkLayer IO Read.Block
+    -> TransactionLayer k 'CredFromKeyK SealedTx
+    -> TimeInterpreter (ExceptT PastHorizonException IO)
+    -> AnyRecentEra
+    -> ChangeAddressGen s
+    -> WalletId
+    -> ExceptT ErrSelectAssets IO DelegationFee
+delegationFee db netLayer txLayer ti era changeAddressGen walletId = do
+    protocolParams <- liftIO $ currentProtocolParameters netLayer
+    WriteTx.withRecentEra era $ \(recentEra :: WriteTx.RecentEra era) -> do
+        feePercentiles <- calculateFeePercentiles $ do
+            (Cardano.Tx (Cardano.TxBody bodyContent) _, _updatedWallet) <-
+                liftIO $ buildTransaction @s @k @n @era
+                    db txLayer ti walletId changeAddressGen protocolParams
+                    defaultTransactionCtx []
+            pure $ case Cardano.txFee bodyContent of
+                Cardano.TxFeeExplicit _ coin -> Fee (fromCardanoLovelace coin)
+                Cardano.TxFeeImplicit Cardano.TxFeesImplicitInByronEra ->
+                    case recentEra of {}
+
+        deposit <- liftIO $ calcMinimumDeposit db netLayer walletId
+        pure DelegationFee { feePercentiles, deposit }
+
+-- | For a given transaction fee estimation calculation evaluates its fee spread
+-- by repeatedly running it (100 times) and recording the results.
+-- In the returned 'FeeSpread', the minimum fee is that which 90% of the sampled
+-- fees are greater than. The maximum fee is the highest fee observed in the
+-- samples.
```

Comment:
`FeeSpread` is now out of date.

And I was going to say this comment now seems redundant with the descriptive types:

> the minimum fee is that which 90% of the sampled fees are greater than. The maximum fee is the highest fee observed in the samples.

but interestingly it claims the second one is the maximum and not the 90th percentile? 😮 I can't tell immediately from the implementation — is this correct?

Reply:
I checked the implementation and it returns 1st and 9th decile (aka 10th and 90th percentile) values; Fixed the comment;